### PR TITLE
Moving service fabrik apiserver to stand alone APIServer with version 1.11

### PR DESCRIPTION
**Pull request by subhankarc**
### What does the PR have?
* Creates a full APIServer
* Authentication via client certs is enabled
* Webhook process for the apiserver CRD validation and defaulting is enabled.

#### Blobs
- new blob **kubernetes-1.11.1/kube-apiserver** is added
#### Src
- **webhooks** go project is added which is a simple REST server which gets callback from the apiserver for defaulting of the resources
#### Packages
- **apiserver** package is added which just copied the kube-apiserver blob
- **webhooks** package is added which builds the webhook go project(part of src)
#### Jobs
- settings.yml.erb of **service-fabrik-broker** and **backup-manager** job is modified to add certificate credentials to connect to apiserver and crds to be added in the runtime.
- **service-fabrik-apiserver** job is modified
  - It now starts the kube-apiserver binary
  - ca, certificate and key are used in order to access the apiserver
  - post_start script is added to check if apiserver is healthy
- **webhook** process is added
  - It does validation and defaulting of apiserver CRDs
  - In the post start of it, mutatingwebhookconfigurations resource is created(to let apiserver know to call the webhook)

Vendor files are generated using `dep` tool, so that can be ignored.

### CODE REVIEW

File: jobs/service-fabrik-apiserver/spec [hunk 1]
```diff
@@ -2,28 +2,46 @@
...
+  admin-password:
+    description: The password for the admin account
+  tls.apiserver.ca:
+    description: "CA Certificate for the apiserver master"
+  tls.apiserver.certificate:
+    description: "Certificate for the apiserver master"
+  tls.apiserver.private_key:
+    description: "Private key for the apiserver master"
+  crds:
+    description: "Private key for the apiserver master"
```
Comment (sauravmndl) [inline jobs/service-fabrik-apiserver/spec]: copy-pasted description from previous :)

File: jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb [hunk 2]
```diff
@@ -1,7 +1,29 @@
...
+  executable: /var/vcap/packages/apiserver/bin/kube-apiserver
+  args:
+  - --anonymous-auth=true
+  - --allow-privileged=false
+  - --apiserver-count=<%= link('service-fabrik-apiserver').instances.size %>
+  - --bind-address=0.0.0.0
+  - --enable-bootstrap-token-auth
+  - --enable-swagger-ui
+  - --etcd-servers=https://<%= p('etcd.url').join(',') %>
+  - --etcd-cafile=/var/vcap/jobs/service-fabrik-apiserver/config/etcd-ca.crt
```
Comment (sauravmndl) [inline jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb]: Can we use variable from job_properties file path **/var/vcap/jobs/service-fabrik-apiserver/config**

File: jobs/service-fabrik-apiserver/spec [hunk 3]
```diff
@@ -2,28 +2,46 @@
...
 templates:
-  bin/service-fabrik-apiserver.erb: bin/service-fabrik-apiserver
+  bin/ensure_apiserver_healthy.erb: bin/ensure_apiserver_healthy
+  bin/post-start.erb: bin/post-start
   config/bpm.yml.erb: config/bpm.yml
   config/etcd-ca.crt.erb: config/etcd-ca.crt
   config/etcd.crt.erb: config/etcd.crt
   config/etcd.key.erb: config/etcd.key
-
+  config/apiserver-key.pem.erb: config/apiserver-key.pem
```
Comment (sauravmndl) [inline jobs/service-fabrik-apiserver/spec]: Can we have name like **apiserver.key , apiserver.crt** in order to keep consistent across system?

File: jobs/service-fabrik-backup-manager/templates/config/settings.yml.erb [hunk 4]
```diff
@@ -54,6 +54,10 @@ production:
   apiserver:
     ip: <%= link("service-fabrik-apiserver").p('ip') %>
     port: <%= link("service-fabrik-apiserver").p('port') %>
+    ca: <%= JSON.dump(link("service-fabrik-apiserver").p('tls.apiserver.ca')) %>
+    certificate: <%= JSON.dump(link("service-fabrik-apiserver").p('tls.apiserver.certificate')) %>
+    private_key: <%= JSON.dump(link("service-fabrik-apiserver").p('tls.apiserver.private_key')) %>
```
Comment (sauravmndl) [inline jobs/service-fabrik-backup-manager/templates/config/settings.yml.erb]: can we have property keys like **cert, key** for consistency ?

File: jobs/service-fabrik-apiserver/spec (hunk 1 above)
Comment (subhankarc) [reply jobs/service-fabrik-apiserver/spec]: :P changed.

File: jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb (hunk 2 above)
Comment (subhankarc) [reply jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb]: There are no jobs_properties file as in bpm I directly run binary and no shell script file is needed. Let me try to declare the variable in the top in this erb.

File: jobs/service-fabrik-apiserver/spec (hunk 3 above)
Comment (subhankarc) [reply jobs/service-fabrik-apiserver/spec]: ok

File: jobs/service-fabrik-backup-manager/templates/config/settings.yml.erb (hunk 4 above)
Comment (subhankarc) [reply jobs/service-fabrik-backup-manager/templates/config/settings.yml.erb]: ok

File: jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb [hunk 5]
```diff
@@ -1,7 +1,29 @@
 ---
 processes:
 - name: service-fabrik-apiserver
-  executable: /var/vcap/jobs/service-fabrik-apiserver/bin/service-fabrik-apiserver
-  persistent_disk: true
-  limits:
-    open_files: 100000
```
Comment (abh1kg) [inline jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb]: Don't we want to keep the open file descriptors limit?

File: src/github.com/cloudfoundry-incubator/webhooks/pkg/webhooks/manager/main.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package main
```
Comment (abh1kg) [inline src/github.com/cloudfoundry-incubator/webhooks/pkg/webhooks/manager/main.go]: Little confused here- would we want to clutter our ephemeral store with these packages and files? Instead, won't it be better to generate the Go executable via compilation or pre-release?
Or has this already been discussed?

File: jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb (hunk 5 above)
Comment (subhankarc) [reply jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb]: bpm internally has a limit of open files. Don't want to override that.

File: src/github.com/cloudfoundry-incubator/webhooks/pkg/webhooks/manager/main.go (hunk 6 above)
Comment (subhankarc) [reply src/github.com/cloudfoundry-incubator/webhooks/pkg/webhooks/manager/main.go]: Webhook job is added for reference. Currently we are not using it. So can be ignored. :)

File: jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb (hunk 5 above)
Comment (abh1kg) [reply jobs/service-fabrik-apiserver/templates/config/bpm.yml.erb]: Ok... fine

File: src/github.com/cloudfoundry-incubator/webhooks/pkg/webhooks/manager/main.go (hunk 6 above)
Comment (abh1kg) [reply src/github.com/cloudfoundry-incubator/webhooks/pkg/webhooks/manager/main.go]: Ok... so, you would remove the webhook job from the packages. fair enough- will approve

**Pull request merged**